### PR TITLE
Fix lyrics Unicode corruption and escaped quotes in Genius plugin

### PR DESCRIPTION
**Pull request by grillofran**
## Problem
The lyrics plugin has two bugs that corrupt fetched lyrics:

1. **Unicode corruption**: Characters like `ò`, `è`, `à` are corrupted to `√≤`, `√®`, etc.
2. **Escaped quotes**: Quotes appear as `\"` instead of `"` in lyrics

### Issue 1: MacRoman encoding misdetection
- **Location**: `RequestHandler.fetch_text()` line 220
- **Cause**: Setting `r.encoding = None` forces requests to use `apparent_encoding`
- **Problem**: For Genius.com (and others), requests incorrectly detects MacRoman instead of UTF-8
- **Result**: UTF-8 bytes `c3 b2` (ò) decoded as MacRoman produces "√≤" (U+221A U+2264)

### Issue 2: Incomplete JSON unescape
- **Location**: `Genius.scrape()` line 576
- **Cause**: The `remove_backslash` regex doesn't handle all escape patterns in JSON
- **Problem**: Genius embeds lyrics in JSON with patterns like `\\"` and `\\\\"` 
- **Result**: After BeautifulSoup processing, escaped quotes remain in final text

### Fix 1: Trust server encoding, fallback to UTF-8
```python
# OLD: r.encoding = None
# NEW:
if not r.encoding:
    r.encoding = 'utf-8'
```
- Respects server's declared encoding (UTF-8 for Genius)
- Falls back to UTF-8 if no encoding specified (safer than apparent_encoding)
- Preserves original intent of handling misconfigured servers

### Fix 2: Iteratively clean escaped quotes
```python
while '\\"' in lyrics:
    lyrics = lyrics.replace('\\"', '"')
```
- Handles variable escape levels (`\"`, `\\\"`, `\\\\\"`)
- Minimal change - keeps original `remove_backslash` regex
- Applied after BeautifulSoup to avoid interfering with HTML parsing

## Testing

Tested with:
- Caparezza - "Argenti Vive" (Italian, many accented characters)
- WestsideGunn - "Heel Cena" (escaped quotes in lyrics)

Before:
```
mi si par√≤ davanti
\\"I got big moves\\"
```

After:
```
mi si parò davanti
"I got big moves"
```

## Impact
- Fixes lyrics for all languages with non-ASCII characters
- Fixes Genius lyrics with quotes
- No breaking changes - maintains backward compatibility
- Minimal code changes (14 lines total)

### CODE REVIEW

**Comment by grillofran** [comment]
Is the 66% patch coverage requirement strict for bug fixes? The changes are small edge-case improvements. Happy to add tests if required!

File: beetsplug/lyrics.py [hunk 1]
```diff
@@ -200,7 +200,11 @@ def get_text(
         url = self.format_url(url, params)
         self.debug("Fetching HTML from {}", url)
         r = self.get(url, **kwargs)
-        r.encoding = None
+        """Trust server's encoding,
```
Comment (semohr) [inline beetsplug/lyrics.py]: We can simplify this:
```
# trust server encoding and default to utf-8
r.encoding = r.encoding or "utf-8"
```

File: beetsplug/lyrics.py [hunk 2]
```diff
@@ -557,7 +561,11 @@ def search(self, artist: str, title: str) -> Iterable[SearchResult]:
     def scrape(cls, html: str) -> str | None:
         if m := cls.LYRICS_IN_JSON_RE.search(html):
             html_text = cls.remove_backslash(m[0]).replace(r"\n", "\n")
-            return cls.get_soup(html_text).get_text().strip()
+            lyrics = cls.get_soup(html_text).get_text().strip()
+            # Clean up any remaining escaped quotes (may need multiple passes)
```
Comment (semohr) [inline beetsplug/lyrics.py]: Comments should be minimal and focus on the why rather than the how:
```
#  Strip HTML and normalize quotes that were left because of JSON...
return re.sub(r'\\+"', '"', lyrics)
```
It should explain why the escaped quotes exist without repeating how the regex works. You have quite some information on this in your PR, maybe we can condense it into a sentence?

**Review by semohr — CHANGES_REQUESTED** [review]
Thank you for the PR. Since this seems to be your first contribution to beets, you might want to have a look at our [developer guides](https://beets.readthedocs.io/en/stable/dev/index.html). Welcome aboard!

> Is the 66% patch coverage requirement strict for bug fixes? The changes are small edge-case improvements. Happy to add tests if required!

While it is not a requirement, for new code tests are always highly appreciated. Here for example showing that JSON encoded lyrics can be parsed as expected would be a good addition.  Same for the inverse case showing that "normal" html lyrics still work as expected would be good (I guess the test already exists, havent look at our testsuit 😁 )

----

More of a general question: how come this was never raised as an issue? Seems like it should be a common issue to me. Have you tried to find related open issues yet?

File: beetsplug/lyrics.py (hunk 1 above)
Comment (snejus) [reply beetsplug/lyrics.py]: @grillofran setting encoding to `None` is required for some websites under `Google` provider, such as **AZLyrics**. Your fix should apply to `Genius` backend only.

You could potentially add a new parameter `reset_encoding: bool=False` to `LyricsRequestHandler.get_text`, and adjust `Google.get_text` accordingly?

**Comment by snejus** [comment]
> Tested with:
> 
>     * Caparezza - "Argenti Vive" (Italian, many accented characters)
> 
>     * WestsideGunn - "Heel Cena" (escaped quotes in lyrics)

Are you happy to add both cases to `test/plugins/lyrics_pages.py`?

**Comment by grillofran** [comment]
Thanks for the feedback! I've updated the PR to address all your points.

- Made the encoding fix Genius-only with a force_utf8 parameter (thanks @snejus)
- Simplified the encoding check to r.encoding = r.encoding or "utf-8" (thanks @semohr)
- Updated the comment to focus on why the escapes exist
- Added test cases for Caparezza (Italian accents) and Arctic Monkeys (quoted lyrics), as the other one was... too explicit

Funny enough.... I couldn't find any issues reporting this in the repo's issue tracker, which I thought was strange given how many people must use Genius. Also, this is my first contribution to beets so apologies if I missed anything. Still learning the ropes! Let me know if there's anything else that needs adjusting.

File: beetsplug/lyrics.py [hunk 3]
```diff
@@ -190,17 +190,23 @@ def format_url(url: str, params: JSONDict | None) -> str:
...
+        By default, trust the server's encoding and requests' apparent_encoding
+        detection. When force_utf8=True, default to UTF-8 if server doesn't
+        specify encoding (avoids MacRoman misdetection on some sites like Genius).
         """
         url = self.format_url(url, params)
         self.debug("Fetching HTML from {}", url)
         r = self.get(url, **kwargs)
-        r.encoding = None
+        if force_utf8:
+            r.encoding = r.encoding or "utf-8"
```
Comment (snejus) [inline beetsplug/lyrics.py:223]: This only handles `Genius` requirements: to clarify, for `Google`, we want:

```py
r.encoding = None
```

same it was before.

**Review by snejus — CHANGES_REQUESTED** [review]

File: beetsplug/lyrics.py (hunk 3 above)
Comment (semohr) [reply beetsplug/lyrics.py:223]: Could we add a test for the specifc case where this is an issue please?

For me (without researching and testing it further) using the encoding of the response seems reasonable.

Comment (snejus) [reply beetsplug/lyrics.py:223]: Luckily, I managed to dig this URL out from my `ipython` history. Add this to `test/plugins/lyrics_pages.py`

```py
    LyricsPage.make(
        "https://www.azlyrics.com/lyrics/derivakat/bountyhunter.html",
        """
        [Romanized:]

        Living legend made of myths and made of stories
        But I'm present, find the mark and I'll take the glory
        Hit the dead-end, cornered like an animal, it just
        Takes one second, I'm nothing if not practical
        But I don't really care if you're right or you're wrong
        Bùguǎn nǐ de wùhuì or you're just a disgrace
        Get in the zone when I put on my song
        Bìrán de xiànshí when I'm up in your face
        Zhēn tài kě qíng kùn zài lóng lǐ de nǐ
        But you'll find in the end that I'll leave you erased
        (If they call?) I'm the one
        (Got a job?) Get it done
        All alone, you and me, I'll be leaving no trace

        No place where I won't go to
        No limits I won't break through
        No godless deed I won't do
        If they got problems, then they know to

        Call the bounty hunter
        Yeah, they got my number
        I'm the problem solver
        Leave you six feet under
        You won't see another
        Winter, spring or summer
        Wǒ huì ràng nǐ xiāoshī
        No way to recover (Hey!)

        Run it up
        Night or day (Hey!)
        Hunt you down
        Run away

        No mercy from this mercenary
        Take away your sanctuary
        Shìlì jiù xiàng 20/20
        Way I see you'll end up buried
        Don't deviate
        Zhǎodào nǐ so I terminate
        Wúlùn, wǒ huì make you pay
        I'll double tap you like bang, bang

        (Dead lilac) A ghost you only heard in rumor
        (Dead lilac) But only if you heard me sooner
        (Dead lilac) Losing your chances and losing your time
        (Dead lilac) shìjiè wèilái méi nǐ de cúnzài (Yeah)

        No place where I won't go to (I won't go to)
        No limits I won't break through (I won't break through)
        No godless deed I won't do (That I won't do)
        If they got problems, then they know to

        Call the bounty hunter (Hey)
        Yeah, they got my number
        I'm the problem solver (Hey)
        Leave you six feet under
        You won't see another (Hey)
        Winter, spring or summer
        Wǒ huì ràng nǐ xiāoshī (Hey)
        No way to recover (Hey!)

        Run it up (Run it up)
        Night or day (Hey!) (Day)
        Hunt you down (Hunt you down)
        Run away

        [English/Chinese:]

        Living legend made of myths and made of stories
        But I'm present, find the mark and I'll take the glory
        Hit the dead-end, cornered like an animal, it just
        Takes one second, I'm nothing if not practical
        But I don't really care if you're right or you're wrong
        不管你的误会 or you're just a disgrace
        Get in the zone when I put on my song
        必然的现实 when I'm up in your face
        真太可情 困在笼里的你
        But you'll find in the end that I'll leave you erased
        (If they call?) I'm the one
        (Got a job?) Get it done
        All alone, you and me, I'll be leaving no trace

        No place where I won't go to
        No limits I won't break through
        No godless deed I won't do
        If they got problems, then they know to

        Call the bounty hunter
        Yeah, they got my number
        I'm the problem solver
        Leave you six feet under
        You won't see another
        Winter, spring or summer
        我会让你消失
        No way to recover (Hey!)

        Run it up
        Night or day (Hey!)
        Hunt you down
        Run away

        No mercy from this mercenary
        Take away your sanctuary
        视力就像 20/20
        Way I see you'll end up buried
        Don't deviate
        找到你 so I terminate
        无论, 我会 make you pay
        I'll double tap you like bang, bang

        (Dead lilac) A ghost you only heard in rumor
        (Dead lilac) But only if you heard me sooner
        (Dead lilac) Losing your chances and losing your time
        (Dead lilac) 世界未来没你的存在 (Yeah)

        No place where I won't go to (I won't go to)
        No limits I won't break through (I won't break through)
        No godless deed I won't do (That I won't do)
        If they got problems, then they know to

        Call the bounty hunter (Hey)
        Yeah, they got my number
        I'm the problem solver (Hey)
        Leave you six feet under
        You won't see another (Hey)
        Winter, spring or summer
        我会让你消失 (Hey)
        No way to recover (Hey!)

        Run it up (Run it up)
        Night or day (Hey!) (Day)
        Hunt you down (Hunt you down)
        Run away
        """,
        artist="Derivakat",
        track_title="Bounty Hunter",
        url_title="Derivakat - Bounty Hunter Lyrics | AZLyrics.com",
        marks=[xfail_on_ci("AZLyrics is blocked by Cloudflare")],
    ),
```

and run

```bash
$ LYRICS_UPDATED=true poe test test/plugins/test_lyrics.py -k azlyrics -vv
```

[asciicast](https://asciinema.org/a/2Qqp0Nd1imH3eemVc6GTivznc)

**Comment by grillofran** [comment]
Hey folks (happy holidays)

Ah, got it. I think I had misunderstood. I added the `else: r.encoding = None` to restore the original behavior for Google/AZLyrics. Should fix the Chinese character issue in the test. Setting `r.encoding = None` explicitly should fix that while keeping the UTF-8 fix for Genius only. Thoughts?

Thanks!

File: beetsplug/lyrics.py [hunk 4]
```diff
@@ -557,7 +574,10 @@ def search(self, artist: str, title: str) -> Iterable[SearchResult]:
     def scrape(cls, html: str) -> str | None:
         if m := cls.LYRICS_IN_JSON_RE.search(html):
             html_text = cls.remove_backslash(m[0]).replace(r"\n", "\n")
-            return cls.get_soup(html_text).get_text().strip()
+            lyrics = cls.get_soup(html_text).get_text().strip()
+            # Genius embeds lyrics in JSON; escape sequences remain after parsing
+            lyrics = re.sub(r'\\+"', '"', lyrics)
+            return lyrics
```
Comment (snejus) [inline beetsplug/lyrics.py]: Now that we've sorted the encoding issue, I wonder whether this could possibly be solved by improving the pattern used by `cls.remove_backslash`?

Comment (snejus) [reply beetsplug/lyrics.py]: @grillofran have you seen my comment by any chance?

**Comment by snejus** [comment]
@semohr worth having a look for you since I made some adjustments under `beetsplug/_utils`

File: beetsplug/_utils/requests.py [hunk 5]
```diff
@@ -21,10 +21,13 @@
 class BeetsHTTPError(requests.exceptions.HTTPError):
     STATUS: ClassVar[HTTPStatus]
 
-    def __init__(self, *args, message: str | None = None, **kwargs) -> None:
-        if not message:
-            message = f"HTTP Error: {self.STATUS.value} {self.STATUS.phrase}"
+    @classmethod
+    def get_status_message(cls) -> str:
```
Comment (semohr) [inline beetsplug/_utils/requests.py]: Is this used yet? Should we postpone add this until it is actually used?

File: beetsplug/_utils/requests.py [hunk 6]
```diff
@@ -71,10 +74,15 @@ class TimeoutAndRetrySession(requests.Session, metaclass=SingletonMeta):
     * raises exceptions for HTTP error status codes
     """
 
-    def __init__(self, *args, **kwargs) -> None:
+    raise_for_status: bool
```
Comment (semohr) [inline beetsplug/_utils/requests.py]: Small comment on what the `raise_for_status` param does would be appreciated.

File: beetsplug/lyrics.py [hunk 7]
```diff
@@ -662,14 +675,37 @@ class Genius(SearchBackend):
     def headers(self) -> dict[str, str]:
         return {"Authorization": f"Bearer {self.config['genius_api_key']}"}
 
+    def create_session(self) -> requests.Session:
+        """Return session for Genius HTTP requests.
+
+        HTTP errors are not raised for Genius API requests, because Genius
```
Comment (semohr) [inline beetsplug/lyrics.py]: This seems a bit like a hacky workaround to me. I would prefer if we catch the `BeetsHTTPError` and get the response from the error if it is needed. That way raising from our session stays consistent.

This should work or am I missing something?
```
try:
    response = requests.get(url)
except BeetsHTTPError as e:
    data = e.response.json()
    print(data)
```

**Review by semohr — CHANGES_REQUESTED** [review]
Thanks for pinging me! 

I don't really think we need to add the `raise_for_status` logic. I do not see the need for it right here, also adding it makes sessions a bit inconsistent imo.

File: beetsplug/_utils/requests.py (hunk 5 above)
Comment (snejus) [reply beetsplug/_utils/requests.py]: It is: see `UnauthorizedMBError` and `TooManyRequestsError`

File: beetsplug/_utils/requests.py (hunk 6 above)
Comment (snejus) [reply beetsplug/_utils/requests.py]: I thought it's pretty self-explanatory?

File: beetsplug/lyrics.py (hunk 7 above)
Comment (snejus) [reply beetsplug/lyrics.py]: This doesn't work since `raise_for_status` raises and `response` does not get returned.

Comment (semohr) [reply beetsplug/lyrics.py]: response should be attached to the error tho since we extend from HTTPError.

Comment (semohr) [reply beetsplug/lyrics.py]: We do something similar in tidal to get response header for the length of the timeout. https://github.com/beetbox/beets/blob/5e6d9adbe79ed8885800bee4804d2eee14f6eb48/beetsplug/tidal/session.py#L88

Comment (snejus) [reply beetsplug/lyrics.py]: Very good point! Adjusted

File: beetsplug/_utils/requests.py (hunk 5 above)
Comment (snejus) [reply beetsplug/_utils/requests.py]: Good point, removed it!

File: beetsplug/lyrics.py [hunk 8]
```diff
@@ -663,13 +676,32 @@ def headers(self) -> dict[str, str]:
         return {"Authorization": f"Bearer {self.config['genius_api_key']}"}
 
     def get_json(self, *args, **kwargs) -> GeniusAPI.Search:
-        response: GeniusAPI.Response = super().get_json(*args, **kwargs)
+        try:
```
Comment (semohr) [inline beetsplug/lyrics.py:679]: Should we move the error handling into the `request` function? Currently I think the `get_text` function does not raise errors on http failures, right?

If it is not necessary, also fine. I see that it was like that before 🙃

Comment (snejus) [reply beetsplug/lyrics.py:679]: > Should we move the error handling into the `request` function? Currently I think the `get_text` function does not raise errors on http failures, right?

Which one? 😅 Doesn't seem like `Genius` defines it?

Comment (semohr) [reply beetsplug/lyrics.py:679]: I mean the logic for the  `TooManyRequestsHTTPError` and `GeniusHTTPError`. Right now the `get_text` method will never raise these. Not sure if it should but I would hope an api is pretty consistent in their error reporting.

We could likely drop the `get_json` override here in favor for a more generic handling in `request` or even in `handle_http_error`.

---

Very much possible that the code here predates the RequestHandler abstraction.

Comment (snejus) [reply beetsplug/lyrics.py:679]: Both of these errors depend on the data present in the JSON response!

Comment (semohr) [reply beetsplug/lyrics.py:679]: > Not sure if it should but I would hope an api is pretty consistent in their error reporting.

We depend on the **error response** containing JSON. If the API uses a consistent error format, though, it’s entirely possible for the response body to contain JSON error data even when you access it via `.text`. We raise the HTTP error before interpreting the response, i.e. before accessing `.text` or `.json()`.

Comment (snejus) [reply beetsplug/lyrics.py:679]: Note we haven't seen the HTML responses fail like that, even the URL root is different here:

```sh
$ beet lyrics 'title:let me love you' -vv -f --no-keep-synced
...
lyrics: Genius: Fetching JSON from https://api.genius.com/search?q=Mario+Let+Me+Love+You
lyrics: Genius: Fetching HTML from https://genius.com/Mario-let-me-love-you-lyrics
...
```

Comment (semohr) [reply beetsplug/lyrics.py:679]: Alright!

**Comment by snejus** [comment]
Any blockers left? @semohr

**Review by semohr — APPROVED** [review]
